README.md
# batik_dcgan

Deep Convolutional Generative Adversarial Network Implementation for Generating Batik Motifs
using Pytorch


changes.txt
Training 28/07/2019
- rmsprop with momentum 0.5
- epoch 10000

Batik_DCGAN.py
import torch

# Check CUDA Available
print(torch.cuda.is_available())
print(torch.version.cuda)

class Generator(nn.Module):
    def __init__(self):
        super(Generator, self).__init__()

        self.init_size = opt.img_size // 4
        self.l1 = nn.Sequential(nn.Linear(opt.latent_dim, 128 * self.init_size ** 2))

        self.conv_blocks = nn.Sequential(
            nn.BatchNorm2d(128),
            nn.Upsample(scale_factor=2),
            nn.Conv2d(128, 128, 3, stride=1, padding=1),
            nn.BatchNorm2d(128, 0.8),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Upsample(scale_factor=2),
            nn.Conv2d(128, 64, 3, stride=1, padding=1),
            nn.BatchNorm2d(64, 0.8),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Conv2d(64, opt.channels, 3, stride=1, padding=1),
            nn.Tanh(),
        )

    def forward(self, z):
        out = self.l1(z)
        out = out.view(out.shape[0], 128, self.init_size, self.init_size)
        img = self.conv_blocks(out)
        return img


class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()

        def discriminator_block(in_filters, out_filters, bn=True):
            block = [nn.Conv2d(in_filters, out_filters, 3, 2, 1), nn.LeakyReLU(0.2, inplace=True), nn.Dropout2d(0.25)]
            if bn:
                block.append(nn.BatchNorm2d(out_filters, 0.8))
            return block

        self.model = nn.Sequential(
            *discriminator_block(opt.channels, 16, bn=False),
            *discriminator_block(16, 32),
            *discriminator_block(32, 64),
            *discriminator_block(64, 128),
        )

        # The height and width of downsampled image
        ds_size = opt.img_size // 2 ** 4
        self.adv_layer = nn.Sequential(nn.Linear(128 * ds_size ** 2, 1), nn.Sigmoid())

    def forward(self, img):
        out = self.model(img)
        out = out.view(out.shape[0], -1)
        validity = self.adv_layer(out)

        return validity

# Loss function
adversarial_loss = torch.nn.BCELoss()

# Initialize generator and discriminator
generator = Generator()
discriminator = Discriminator()

if cuda:
    generator.cuda()
    discriminator.cuda()
    adversarial_loss.cuda()

# Initialize weights
generator.apply(weights_init_normal)
discriminator.apply(weights_init_normal)

# Configure data loader
os.makedirs("../../data/mnist", exist_ok=True)
dataloader = torch.utils.data.DataLoader(
    datasets.MNIST(
        "../../data/mnist",
        train=True,
        download=True,
        transform=transforms.Compose(
            [transforms.Resize(opt.img_size), transforms.ToTensor(), transforms.Normalize([0.5], [0.5])]
        ),
    ),
    batch_size=opt.batch_size,
    shuffle=True,
)

# Optimizers
optimizer_G = torch.optim.Adam(generator.parameters(), lr=opt.lr, betas=(opt.b1, opt.b2))
optimizer_D = torch.optim.Adam(discriminator.parameters(), lr=opt.lr, betas=(opt.b1, opt.b2))

Tensor = torch.cuda.FloatTensor if cuda else torch.FloatTensor

# ----------
#  Training
# ----------

for epoch in range(opt.n_epochs):
    for i, (imgs, _) in enumerate(dataloader):

        # Adversarial ground truths
        valid = Variable(Tensor(imgs.shape[0], 1).fill_(1.0), requires_grad=False)
        fake = Variable(Tensor(imgs.shape[0], 1).fill_(0.0), requires_grad=False)

        # Configure input
        real_imgs = Variable(imgs.type(Tensor))

        # -----------------
        #  Train Generator
        # -----------------

        optimizer_G.zero_grad()

        # Sample noise as generator input
        z = Variable(Tensor(np.random.normal(0, 1, (imgs.shape[0], opt.latent_dim))))

        # Generate a batch of images
        gen_imgs = generator(z)

        # Loss measures generator's ability to fool the discriminator
        g_loss = adversarial_loss(discriminator(gen_imgs), valid)

        g_loss.backward()
        optimizer_G.step()

        # ---------------------
        #  Train Discriminator
        # ---------------------

        optimizer_D.zero_grad()

        # Measure discriminator's ability to classify real from generated samples
        real_loss = adversarial_loss(discriminator(real_imgs), valid)
        fake_loss = adversarial_loss(discriminator(gen_imgs.detach()), fake)
        d_loss = (real_loss + fake_loss) / 2

        d_loss.backward()
        optimizer_D.step()

        print(
            "[Epoch %d/%d] [Batch %d/%d] [D loss: %f] [G loss: %f]"
            % (epoch, opt.n_epochs, i, len(dataloader), d_loss.item(), g_loss.item())
        )

        batches_done = epoch * len(dataloader) + i
        if batches_done % opt.sample_interval == 0:
            save_image(gen_imgs.data[:25], "images/%d.png" % batches_done, nrow=5, normalize=True)

main.py
from torch import nn
import torch
from torchvision import datasets
from torchvision import transforms
from torch.autograd import Variable
import numpy as np
from torchvision.utils import save_image
import torchvision.utils as vutils
import torch.nn.parallel
import torch.optim as optim


# variables

BATCH_SIZE = 128
DATASET_ROOT_DIR = r'F:\miqdude\Batik'
# 'C:\Users\CORE\Desktop\miqdude\Kawung'
LEARNING_RATE = 0.002
EPOCH = 10000
SAVE_INTERVAL = 100
NOISE_VECTOR_DIM = 100
FEATURE_MAPS = 64
CHANNEL_SIZE = 3        # image channel RGB

# check CUDA Support GPU
cuda = True if torch.cuda.is_available() else False


# initiate weight at class declaration
def weights_init_normal(m):
    classname = m.__class__.__name__
    if classname.find("Conv") != -1:
        torch.nn.init.normal_(m.weight.data, 0.0, 0.02)
    elif classname.find("BatchNorm2d") != -1:
        torch.nn.init.normal_(m.weight.data, 1.0, 0.02)
        torch.nn.init.constant_(m.bias.data, 0.0)



class Generator(nn.Module):
    def __init__(self):
        super(Generator, self).__init__()

        """ 
        Inverse transpose convolution network
        tuned from DCGAN paper
        """
        self.main = nn.Sequential(
            nn.ConvTranspose2d(NOISE_VECTOR_DIM, FEATURE_MAPS * 8, 4, 1, 0, bias = False),
            nn.BatchNorm2d(FEATURE_MAPS * 8),
            nn.ReLU(True),
            nn.ConvTranspose2d(FEATURE_MAPS * 8, FEATURE_MAPS * 4, 4, 2, 1, bias = False),
            nn.BatchNorm2d(FEATURE_MAPS * 4),
            nn.ReLU(True),
            nn.ConvTranspose2d(FEATURE_MAPS * 4, FEATURE_MAPS * 2, 4, 2, 1, bias = False),
            nn.BatchNorm2d(FEATURE_MAPS * 2),
            nn.ReLU(True),
            nn.ConvTranspose2d(FEATURE_MAPS * 2, FEATURE_MAPS, 4, 2, 1, bias = False),
            nn.BatchNorm2d(FEATURE_MAPS),
            nn.ReLU(True),
            nn.ConvTranspose2d(FEATURE_MAPS, CHANNEL_SIZE, 4, 2, 1, bias = False),
            nn.Tanh()
        )

    # function called inside __init__
    def forward(self, input):
        output = self.main(input)
        return output


class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()

        """ 
        convolution network tuned   
        according to paper DCGAN by Alec Radford
        """
        self.main = nn.Sequential(
            nn.Conv2d(CHANNEL_SIZE, FEATURE_MAPS, 4, 2, 1, bias = False),
            nn.LeakyReLU(0.2, inplace = True),
            nn.Conv2d(FEATURE_MAPS, FEATURE_MAPS * 2, 4, 2, 1, bias = False),
            nn.BatchNorm2d(FEATURE_MAPS * 2),
            nn.LeakyReLU(0.2, inplace = True),
            nn.Conv2d(FEATURE_MAPS * 2, FEATURE_MAPS * 4, 4, 2, 1, bias = False),
            nn.BatchNorm2d(FEATURE_MAPS * 4),
            nn.LeakyReLU(0.2, inplace = True),
            nn.Conv2d(FEATURE_MAPS * 4, FEATURE_MAPS * 8, 4, 2, 1, bias = False),
            nn.BatchNorm2d(FEATURE_MAPS * 8),
            nn.LeakyReLU(0.2, inplace = True),
            nn.Conv2d(FEATURE_MAPS * 8, 1, 4, 1, 0, bias = False),
            nn.Sigmoid()
        )

    def forward(self, input):
        output = self.main(input)
        return output.view(-1)
        

# Loss function
adversarial_loss = torch.nn.BCELoss()

# Initialize generator and discriminator
generator = Generator()
discriminator = Discriminator()

# enable cuda compute on generator, discriminator, and loss function
if cuda:
    generator.cuda()
    discriminator.cuda()
    adversarial_loss.cuda()

# Initialize weights
generator.apply(weights_init_normal)
discriminator.apply(weights_init_normal)

# Configure data loader for feeding data to discriminator every batch
dataloader = torch.utils.data.DataLoader(
    datasets.ImageFolder(
        DATASET_ROOT_DIR,
        transform = transforms.Compose(
            [transforms.Resize((64,64)), transforms.ToTensor(), transforms.Normalize([0.5], [0.5])]
        )
    ),
    batch_size=BATCH_SIZE,
    shuffle=True,
)

# # Optimizers using Adam
# optimizerG = torch.optim.Adam(generator.parameters(), lr=LEARNING_RATE, betas=(0.5, 0.999))
# optimizerD = torch.optim.Adam(discriminator.parameters(), lr=LEARNING_RATE, betas=(0.5, 0.999))

# Optimizers using RMSProp
optimizerG = torch.optim.SGD(generator.parameters(), lr=LEARNING_RATE, momentum =0.5)
optimizerD = torch.optim.SGD(discriminator.parameters(), lr=LEARNING_RATE, momentum=0.5)


# Tensor object for gpu compute
Tensor = torch.cuda.FloatTensor if cuda else torch.FloatTensor

print("Training Start.....")

# The training process
for epoch in range(EPOCH):

    for i, data in enumerate(dataloader, 0):
        
        # Train discriminator with real images
        discriminator.zero_grad() # clears the gradient
        
        real, _ = data
        input = Variable(real.type(Tensor))  # input images must be Tensor for gpu compute
        target = Variable(torch.ones(input.size()[0]).cuda())
        output = discriminator(input)
        errD_real = adversarial_loss(output, target)
        

        # Train discriminator with fake images
        # Generator generate fake images
        noise = Variable(torch.randn(input.size()[0], 100, 1, 1).cuda())
        fake = generator(noise)
        target = Variable(torch.zeros(input.size()[0]).cuda())
        
    
        output = discriminator(fake.detach()).view(-1)
        errD_fake = adversarial_loss(output, target)

        # calculate error/loss on discriminator    
        errD = errD_real + errD_fake
        errD.backward() # backward propagation
        optimizerD.step() # optimizing model

        # Train generator
        generator.zero_grad()
        target = Variable(torch.ones(input.size()[0]).cuda())
        output = discriminator(fake)
        errG = adversarial_loss(output, target)
        errG.backward()
        optimizerG.step()
        
        # print training logs
        print('[%d/%d][%d/%d] Loss_D: %.4f Loss_G: %.4f' % (epoch, EPOCH, i, len(dataloader), errD.data, errG.data))
        
        # logging training loss
        with open("training_log.txt", "a") as myfile:
            myfile.write('[%d/%d][%d/%d] Loss_D: %.4f Loss_G: %.4f\n' % (epoch, EPOCH, i, len(dataloader), errD.data, errG.data))
            myfile.close()

        # Save images and models per interval and last epoch
        if (epoch % SAVE_INTERVAL) == 0 or (epoch == EPOCH - 1):
            vutils.save_image(real, '%s/real_samples_%03d.png' % ("./results",epoch), normalize = True)
            fake = generator(noise)
            vutils.save_image(fake.data, '%s/fake_samples_epoch_%03d.png' % ("./results", epoch), normalize = True)

        # save models once very 1000th epoch and last epoch
        if (epoch % 2000 == 0) or (epoch == EPOCH -1):  
            # save models for future use
            # save generator 
            torch.save({
                'epoch': epoch,
                'model_state_dict': generator.state_dict(),
                'optimizer_state_dict': optimizerG.state_dict(),
                'loss': errG,
            }, './models/generator_%03d.pt' % epoch)

            # save discriminator 
            torch.save({
                'epoch': epoch,
                'model_state_dict': discriminator.state_dict(),
                'optimizer_state_dict': optimizerD.state_dict(),
                'loss': errD,
            }, './models/discriminator_%03d.pt' % epoch)